### PR TITLE
i#5320: Fix error handling in in x86 decode_modrm

### PR DESCRIPTION
**Pull request by dvyukov**
Fix the following error reported during fuzzing:

core/ir/x86/decode.c:1758:36: runtime error: division of
    -2147483648 by -1 cannot be represented in type 'int'
  in decode_modrm core/ir/x86/decode.c:1758:36
  in decode_operand core/ir/x86/decode.c
  in decode_common core/ir/x86/decode.c:2574:18
  in decode core/ir/x86/decode.c:2695:12

Issues: #5320

### CODE REVIEW

**Comment by dvyukov** [comment]
Where should I add a test for this?
The closest I found is this:
https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/api/ir_x86.c#L1588-L1614
but it looks slightly different. Should I add a new test in this file that tests completely invalid instructions.

**Comment by derekbruening** [comment]
> Where should I add a test for this? The closest I found is this: https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/api/ir_x86.c#L1588-L1614 but it looks slightly different. Should I add a new test in this file that tests completely invalid instructions.

The existing x86 IR tests:
+ construction of every opcode in suite/tests/api/ir_x86_*.h
+ decode random binary blob suite/tests/api/dis-udis86-randtest.raw
+ binutils tests third_party/binutils/test_decenc/
+ ad hoc manual tests suite/tests/api/ir_x86.c

Since the udis86 test isn't meant to be appended to, for a small number of invalid encoding spot-checks ir_x86.c makes the most sense, in a new function.  There are also some invalid cases elsewhere in the file like test_vsib() `b2`.

**Comment by dvyukov** [comment]
I've added a test.
Most builders succeeded, one still failed:
```
2022-02-05T10:18:45.4135096Z 286/369 Test #286: code_api|tool.drcachesim.scattergather ...........................***Failed  Required regular expression not found. Regex=[^AVX-512 gather ok
```
I've fixed a number of other failure, but don't see how this one is related.
Can it be a flake?

**Comment by derekbruening** [comment]
> ```
> 2022-02-05T10:18:45.4135096Z 286/369 Test #286: code_api|tool.drcachesim.scattergather ...........................***Failed  Required regular expression not found. Regex=[^AVX-512 gather ok
> ```
> 
> I've fixed a number of other failure, but don't see how this one is related. Can it be a flake?

Never seen that before:
```
ASSERT FAILURE: /home/runner/work/dynamorio/dynamorio/clients/drcachesim/tracer/tracer.cpp:394: towrite <= ipc_pipe.get_atomic_write_size() && towrite > 0 ()
```

It does seem unlikely to come from this PR (we'll see on the update re-run).  Filed #5329; hopefully it will repro somewhere else and not remain an unsolvable mystery.  @abhinav92003 you add this test recently: any ideas?

File: suite/tests/api/ir_x86.c [hunk 1]
```diff
@@ -1028,6 +1028,21 @@ test_modrm16(void *dc)
     test_modrm16_helper(dc, DR_REG_BX, DR_REG_NULL, 0x80, 5);
 }
 
+/* Just check that decoding fails silently on invalid data. */
+static void
+test_modrm_invalid(void *dc)
+{
+/* Decoding succeeds on 32-bits. */
+#ifdef X64
+    /* Fuzzer-generated random data (i#5320). */
```
Comment (derekbruening) [inline suite/tests/api/ir_x86.c:1037]: nit: s/i5320/i#5320/.  I went ahead and did this to expedite merging.

**Review by derekbruening — APPROVED** [review]
Thank you for the bug report and fix.

**Comment by abhinav92003** [comment]
> @abhinav92003 you add this test recently: any ideas?

Apologies, I think this ping got lost in my inbox.

> ASSERT FAILURE: /home/runner/work/dynamorio/dynamorio/clients/drcachesim/tracer/tracer.cpp:394: towrite <= ipc_pipe.get_atomic_write_size() && towrite > 0 ()

I haven't seen this before either. Thanks for creating the issue. We can discuss further there.

**Pull request merged**